lib/execution_engine2/utils/clients.py

```diff
@@ -0,0 +1,112 @@
+""" Contains the various clients EE2 needs to communicate with other services it depends on. """
+
+# Note on testing - this class is not generally unit-testable, and is only tested fully in
+# integration tests.
+
+from typing import Dict
+
+from execution_engine2.authorization.roles import AdminAuthUtil
+from execution_engine2.authorization.workspaceauth import WorkspaceAuth
+from execution_engine2.utils.CatalogUtils import CatalogUtils
+from execution_engine2.utils.Condor import Condor
+from execution_engine2.sdk.EE2Constants import ADMIN_READ_ROLE, ADMIN_WRITE_ROLE
+from execution_engine2.utils.KafkaUtils import KafkaClient
+from execution_engine2.utils.SlackUtils import SlackClient
+
+from installed_clients.authclient import KBaseAuth
+from installed_clients.WorkspaceClient import Workspace
+
+
+class UserClientSet:
+    """
+    Clients required by EE2 for communicating with other services that need to be instantiated
+    on a per user basis. Also contains the user credentials for ease of use.
+    """
+
+    def __init__(self, cfg: Dict[str, str], user_id: str, token: str):
+        """
+        Initialize the client set.
+
+        cfg - the configuration dictionary
+        user_id - the ID of the user to be used to initialize the client set.
+        token - the token of the user to be used to initialize the client set. Note that the set
+            trusts that the token actually belongs to the user ID, and currently does not
+            independently check the validity of the user ID.
+
+        Expected keys in config:
+        workspace-url - the URL of the kbase workspace service
+        """
+        if not user_id or not user_id.strip():
+            raise ValueError("user_id is required")
+        if not token or not token.strip():
+            raise ValueError("token is required")
+        if not cfg:
+            raise ValueError("cfg is required")
+        self._user_id = user_id
+        self._token = token
+
+        # Do a check that the url actually points to the workspace?
+        # Also maybe consider passing in the workspace url rather than the dict, but the ClientSet
+        # below will need lots of params so a dict makes sense there, maybe keep the apis similar?
+        # TODO the client throws a 'X is not a valid url' error if the url isn't valid, improve
+        #      by catching & rethrowing with a more clear message that the config is wrong
+        ws_url = cfg.get("workspace-url")  # may want to make the keys constants?
+        if not ws_url or not ws_url.strip():
+            raise ValueError("missing workspace-url in configuration")
+        self._workspace = Workspace(ws_url, token=token)
+        self._workspace_auth = WorkspaceAuth(user_id, self._workspace)
+
+    # create_autospec can't mock instance variables, so we add some boilerplate
+    # Could make these properties but then they return MagicMocks which don't update with API
+    # changes
+
+    def user_id(self):
```

Comment:
This seems  strange to do it this way. Are other kbase codebases doing it this way?

Reply:
I don't think so. You only need to do this when you have a mock where you need to mock the instance variables, which doesn't happen all that often in my experience. 

A better way to do this might be to make a constructor that takes a workspace client, user id, and token, and then you can just mock the workspace client and you don't need to mock the client set. I think I'll do that, then this becomes a non-issue.

Reply:
Did that, this is now moot, no more getters